[exporter][prometheusremotewrite] Additional metrics for the exporter

gotjosh opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

Often, we have to do anylsis of the current state of the exporter which includes the current level of concurrency and throughput.

For the concurrency, we need to rely on revising each individual configuration which is time consuming, here I'm proposing to have it as a metric that we can keep a record of and see it through observability.

For the throughput, while tecnically `otelcol_processor_batch_batch_send_size_count` exists, this is not divisible by type (e.g. metrics vs traces) and is not specific to exporter. Whilst I'm also interested in splitting `batch_send_size_count` by type (because that also includes OTLP batches and this metric is only for RW), I think this is part of a separate conversation and it doesn't hurt to have the single counter that's specific for remote write.

~~There's one more thing here (and please advise if you'd rather see a different PR for it) - I'm also adding the `url` as attribute to _ALL_ the metrics  as I think this is key to identify where the exporter is pointing to.~~

This is tecnically my first contribution to the repo so I'd appreciate any guidiance on what else can I need to do. Telemetry tests appear to be auto generated so I didn't add any to this PR but please let me know otherwise.


<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue

This technically a prequite for the work that I'm about to do as part of https://github.com/open-telemetry/opentelemetry-collector/issues/12255 but is not directly related and can go individually.

Signed-off-by: gotjosh <josue.abreu@gmail.com>

gotjosh committed:
[exporter][prometheusremotewrite] Additional metrics for the exporter

Often, we have to do anylsis of the current state of the exporter which includes the current level of concurrency and throughput.

For the concurrency, we need to rely on revising each individual configuration which is time consuming, here I'm proposing to have it as a metric that we can keep a record of and see it through observability.

For the throughput, while tecnically `otelcol_processor_batch_batch_send_size_count` exists, this is not divisible by type (e.g. metrics vs traces) and is not specific to exporter. Whilst I'm also interested in splitting `batch_send_size_count` by type (because that also includes OTLP batches and this metric is only for RW), I think this is part of a separate conversation and it doesn't hurt to have the single counter that's specific for remote write.

There's one more thing here (and please advise if you'd rather see a different PR for it) - I'm also adding the `url` as attribute to _ALL_ the metrics  as I think this is key to identify where the exporter is pointing to.

This is tecnically my first contribution to the repo so I'd appreciate any guidiance on what else can I need to do. Telemetry tests appear to be auto generated so I didn't add any to this PR but please let me know otherwise.

Signed-off-by: gotjosh <josue.abreu@gmail.com>

ArthurSens commented on exporter/prometheusremotewriteexporter/exporter.go:
```diff
@@ -101,6 +111,7 @@ func newPRWTelemetry(set exporter.Settings) (prwTelemetry, error) {
 		telemetryBuilder: telemetryBuilder,
 		otelAttrs: []attribute.KeyValue{
 			attribute.String("exporter", set.ID.String()),
+			attribute.String("endpoint", endpointURL.String()),
```
If we add this attribute to otelAttrs, will this be added to all metrics of the exporter? Or just the new ones you're adding in the PR?

gotjosh replied:
Depends on what you mean with "all metrics of the exporter".

It'll certainly be added to all metrics defined specifically for the exporter but not the generic ones, in this case we'll add them to:

- `otelcol_exporter_prometheusremotewrite_consumers`
- `otelcol_exporter_prometheusremotewrite_sent_batch_count`
- `otelcol_exporter_prometheusremotewrite_translated_time_series`
- `otelcol_exporter_prometheusremotewrite_failed_translations`

but **not** to other metrics such as`otelcol_exporter_queue_capacity` or `otelcol_exporter_send_failed_metric_points`

gotjosh replied:
<img width="1436" alt="image" src="https://github.com/user-attachments/assets/b2208a4a-0ad8-4a0b-839b-ef7999b39c5e" />

ArthurSens reviewed (commented):
I had a super quick look, I'll do proper testing and review tomorrow but I had one doubt already

gotjosh committed:
remove the count suffix

the exporter automatically adds `total` at the end so I think this is good enough.

gotjosh commented:
I think this is now ready for reviewed, I have manually tested my changes and they look from my side.

Does this need a changelog entry? I couldn't tell if we do changelog entries for new metrics.
<img width="1701" alt="Screenshot 2025-02-14 at 14 16 03" src="https://github.com/user-attachments/assets/0d59858b-93be-409d-8751-e75fa56cf41b" />
<img width="1682" alt="Screenshot 2025-02-14 at 14 15 12" src="https://github.com/user-attachments/assets/764e382d-cb2e-48b8-b4f5-c4c41d87a77f" />

gotjosh commented on exporter/prometheusremotewriteexporter/exporter.go:
```diff
@@ -117,7 +128,7 @@ func newPRWExporter(cfg *Config, set exporter.Settings) (*prwExporter, error) {
 		return nil, errors.New("invalid endpoint")
 	}
 
-	prwTelemetry, err := newPRWTelemetry(set)
+	telemetry, err := newPRWTelemetry(set, endpointURL)
```
renamed this variable because it was shadowing a package name.

ArthurSens commented on exporter/prometheusremotewriteexporter/exporter.go:
```diff
@@ -38,13 +38,23 @@ import (
+}
+
+func (p *prwTelemetryOtel) recordRemoteWriteSentBatch(ctx context.Context) {
+	p.telemetryBuilder.ExporterPrometheusremotewriteSentBatchCount.Add(ctx, 1, metric.WithAttributes(p.otelAttrs...))
```
```suggestion
	p.telemetryBuilder. ExporterPrometheusremotewriteSentBatch.Add(ctx, 1, metric.WithAttributes(p.otelAttrs...))
```

That's the name you've added in `telemetryBuilder` 😛

gotjosh replied:
Done in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37912/commits/f535f8e580206b4f7bf622abd8311b8d09e7716b

ArthurSens reviewed (commented):
I think we should add the remoteWriteURL as an attribute in a different PR, it's very easy to split and we can easily rollback PRs in case something unrelated goes wrong.

Could someone with permissions trigger the CI?

gotjosh committed:
Remove the attribute and fix metric name

Signed-off-by: gotjosh <josue.abreu@gmail.com>

gotjosh commented:
> I think we should add the remoteWriteURL as an attribute in a different PR, it's very easy to split and we can easily rollback PRs in case something unrelated goes wrong.

Done in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37912/commits/f535f8e580206b4f7bf622abd8311b8d09e7716b

gotjosh committed:
Remove the endpointURL from the arguments to telemetry

We don't need it anymore

Signed-off-by: gotjosh <josue.abreu@gmail.com>

gotjosh committed:
add the telemetry to the test

Signed-off-by: gotjosh <josue.abreu@gmail.com>

gotjosh committed:
Merge branch 'main' into additional-metrics-for-remote-write-exporter

gotjosh commented:
Thanks @ArthurSens, I would appreciate if anyone could please kick in the test again - I believe I have addressed everything now that I have figured out how to run the test / linters locally.

gotjosh committed:
Add a changelog entry

Signed-off-by: gotjosh <josue.abreu@gmail.com>

ArthurSens reviewed (approved):
LGTM

dashpole commented on exporter/prometheusremotewriteexporter/metadata.yaml:
```diff
@@ -27,3 +27,17 @@ telemetry:
+    exporter_prometheusremotewrite_consumers:
+      enabled: true
+      description: Number of configured workers to use to fan out the outgoing requests
+      unit: "1"
```
```suggestion
      unit: "{consumer}"
```

Or {worker} could work as well

gotjosh replied:
Addressed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37912/commits/62bcf74b1d50eb487fe26481f73b173b42e1d15d

dashpole commented on exporter/prometheusremotewriteexporter/metadata.yaml:
```diff
@@ -27,3 +27,17 @@ telemetry:
+      enabled: true
+      description: Number of configured workers to use to fan out the outgoing requests
+      unit: "1"
+      gauge:
```
This needs to be an updown counter

ArthurSens replied:
real question, What's the difference between gauge and updown counter?

gotjosh replied:
I had the same question and ended up [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#gauge).

I think the answer is: if the `sum` of values across a dimension makes sense then it should be an up and down counter.  In this case, it might make sense to sum the number of consumers across several collectors sending to the same endpoint to try and identify what kind of concurrency you have against your metric storage so I think it makes sense.

gotjosh replied:
@dashpole I have adressed your feedback in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37912/commits/62bcf74b1d50eb487fe26481f73b173b42e1d15d

dashpole replied:
It mostly has to do with how they are aggregated by defalt.  UpDownCounters aggregate by summing.  Gauges aggregate by averaging (or computing other things about the distribution).

An example of an UpDownCounter is memory usage.  You could reasonably measure total memory usage across a number of pods by summing. But you would never want to sum a temperature, so that is a gauge.

ArthurSens replied:
Thank you both for the explanation ❤️

gotjosh committed:
address PR feedback from david

Signed-off-by: gotjosh <josue.abreu@gmail.com>

gotjosh commented on exporter/prometheusremotewriteexporter/metadata.yaml:
```diff
@@ -27,3 +27,17 @@ telemetry:
+    exporter_prometheusremotewrite_sent_batch:
+      enabled: true
+      description: Number of remote write request batches sent to the remote write endpoint regardless of success or failure
+      unit: "1"
```
Addressed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37912/commits/62bcf74b1d50eb487fe26481f73b173b42e1d15d

dashpole commented on exporter/prometheusremotewriteexporter/documentation.md:
```diff
@@ -14,6 +22,14 @@ Number of translation operations that failed to translate metrics from Otel to P
 | ---- | ----------- | ---------- | --------- |
 | 1 | Sum | Int | true |
 
+### otelcol_exporter_prometheusremotewrite_sent_batch
```
Based on https://github.com/open-telemetry/semantic-conventions/blob/e70853397b73ca9cb85b1df45182b6a3e299c85e/docs/general/naming.md#naming-rules-for-counters-and-updowncounters, it seems like this should be `batches`, rather than `batch`

gotjosh replied:
Should I also re-add count which I had previously removed thinking it was redudant?

dashpole replied:
This is a Counter, not an updowncounter, so the updowncounter guidance does not apply here

> Metric names SHOULD NOT be pluralized, unless the value being recorded represents discrete instances of a [countable quantity](https://wikipedia.org/wiki/Count_noun). Generally, the name SHOULD be pluralized only if the unit of the metric in question is a non-unit (like {fault} or {operation}).

This is a count of discrete instances of a countable quantity, so it should be plural.

gotjosh replied:
Done in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37912/commits/dd3c4665b21c07787e25cd573336fd2ed1d8c627

Appreciate the feedback - can you please kick off the CI again?

gotjosh committed:
counters should end with a pluralization

Signed-off-by: gotjosh <josue.abreu@gmail.com>

MovieStoreGuy committed:
Merge branch 'main' into additional-metrics-for-remote-write-exporter

jmichalek132 commented:
LGTM :)

jmichalek132 reviewed (approved):
LGTM.

gotjosh committed:
address review feedback

Signed-off-by: gotjosh <josue.abreu@gmail.com>

gotjosh commented on exporter/prometheusremotewriteexporter/metadata.yaml:
```diff
@@ -27,3 +27,17 @@ telemetry:
+      enabled: true
+      description: Number of configured workers to use to fan out the outgoing requests
+      unit: "{consumer}"
+      sum:
```
I thought the same 😄 , but @dashpole and I had a conversation about it: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37912#discussion_r1961932230

Let me know what you think.

dehaansa replied:
I think the part that was mostly throwing me off was using "Add" to set the value, but I see that it's only set on exporter initialization, so functionally it should be fine. 👍

gotjosh commented on .chloggen/additional-metrics-for-remote-write-exporter.yaml:
```diff
@@ -0,0 +1,30 @@
+note: |
+    Adds additional metrics to the Prometheus Remote Write Exporter. The new metrics are:
+    - `otelcol_exporter_prometeusremotewrite_sent_batch`: The total number of batches sent to the remote write endpoint.
+    - `otelcol_exporter_prometeusremotewrite_consumers`: The total number of consumers that are currently sending data to the remote write endpoint.
```
Done in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37912/commits/36acb443227a1c8f2f6825657b9094ab5c6b5da2, thanks!

gotjosh commented on .chloggen/additional-metrics-for-remote-write-exporter.yaml:
```diff
@@ -0,0 +1,30 @@
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: |
+    Adds additional metrics to the Prometheus Remote Write Exporter. The new metrics are:
+    - `otelcol_exporter_prometeusremotewrite_sent_batch`: The total number of batches sent to the remote write endpoint.
```
Done in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37912/commits/36acb443227a1c8f2f6825657b9094ab5c6b5da2, thanks!

andrzej-stencel committed:
Merge branch 'main' into additional-metrics-for-remote-write-exporter
